nullaway/src/test/java/com/uber/nullaway/NullAwayContractsTests.java

```diff
@@ -267,4 +267,161 @@ public void customContractAnnotation() {
+            "    // null, _ -> true",
+            "    if (o1 == null) {",
+            "      return true;",
+            "    }",
+            "    // _, null -> true",
+            "    if (o2 == null) {",
+            "      return true;",
+            "    }",
+            "    // Function cannot declare a contract for false",
+            "    return System.currentTimeMillis() % 100 == 0;",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test1(@Nullable Object o) {",
+            "    return NullnessChecker.eitherIsNullOrRandom(\"\", o)",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      ? o.toString()",
+            "      : o.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractDeclaringNullOrRandomFalse() {
+    makeTestHelperWithArgs(
+            Arrays.asList(
+                "-d",
+                temporaryFolder.getRoot().getAbsolutePath(),
+                "-XepOpt:NullAway:AnnotatedPackages=com.uber"))
+        .addSourceLines(
+            "NullnessChecker.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "public class NullnessChecker {",
+            "  @Contract(\"null -> false\")",
+            "  static boolean isHashCodeZero(@Nullable Object o) {",
+            "    // null -> false",
+            "    if (o == null) {",
+            "      return false;",
+            "    }",
+            "    // Function cannot declare a contract for true",
+            "    return o.hashCode() == 0;",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test1(@Nullable Object o) {",
+            "    return NullnessChecker.isHashCodeZero(o)",
+            "      ? o.toString()",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      : o.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePath() {
+    makeTestHelperWithArgs(
+            Arrays.asList(
+                "-d",
+                temporaryFolder.getRoot().getAbsolutePath(),
+                "-XepOpt:NullAway:AnnotatedPackages=com.uber"))
+        .addSourceLines(
+            "NullnessChecker.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "public class NullnessChecker {",
+            "  @Contract(\"!null -> false\")",
```

Comment:
So, I still think this is an example of why updating on `"!null -> true|false"` is not very useful. Here, either way, `required` is non-null. Is there a test case or example where we would want a different answer for a method with `@Contract(\"!null -> false\")` vs one with no contract annotation, from the part of the caller? (Separate for there being a case for validating this contract within the body of the callee, of course!)

Reply:
I agree that what it's not very useful, I added this test case in order to prove that we don't declare a `NONNULL` argument `NULLABLE` and incorrectly describe impossible dereferences.